onnx/defs/nn/defs.cc

```diff
@@ -2466,24 +2466,21 @@ ONNX_OPERATOR_SET_SCHEMA(
              "tensor(double)",
              "tensor(bfloat16)"},
             "Constrain input and output types to all numeric tensors.")
-        .FunctionBody(FunctionBodyHelper::BuildNodes(
-            {// nodes: {outputs, op, inputs, attributes}
-             FunctionBodyHelper::Const<float>("Exponent", 2.0f),
-             FunctionBodyHelper::Const<float>("Epsilon", float(1e-9)),
-             {{"X_RM"},
-              "ReduceMean",
-              {"X"},
-              {MakeRefAttribute("axes", AttributeProto::INTS)}},
-             {{"EX_squared"}, "Pow", {"X_RM", "Exponent"}},
-             {{"X_squared"}, "Pow", {"X", "Exponent"}},
-             {{"E_Xsquared"},
-              "ReduceMean",
-              {"X_squared"},
-              {MakeRefAttribute("axes", AttributeProto::INTS)}},
-             {{"Variance"}, "Sub", {"E_Xsquared", "EX_squared"}},
-             {{"STD"}, "Sqrt", {"Variance"}},
-             {{"X_variance"}, "Sub", {"X", "X_RM"}},
-             {{"Processed_STD"}, "Add", {"STD", "Epsilon"}},
-             {{"Y"}, "Div", {"X_variance", "Processed_STD"}}})));
+        .FunctionBody(R"ONNX(
+        {
+          Exponent = Constant <value = float {2.0}>()
+          Epsilon = Constant <value = float {0.000000001}>()
+          X_RM = ReduceMean <axes : ints = @axes> (X)
```

Comment:
I think we have a limitation in current code which requires function attr name to match node attr name.... can we get rid of this limitation now? If we change the attr name for function from axes to something different will it work?

Reply:
I don't think there was a limitation in proto. The builder utility earlier had a limitation (that is, it had only one name parameter that it used for both). Even that was subsequently removed. See https://github.com/onnx/onnx/blob/a5e7ee51176bf78a60c118758174e13d85a87b46/onnx/defs/attr_proto_util.h#L45 and https://github.com/onnx/onnx/blob/a5e7ee51176bf78a60c118758174e13d85a87b46/onnx/defs/attr_proto_util.h#L54 ... unless I misunderstood your question.

Reply:
On a different note: I don't think the "type" field is required in an attribute-reference ... however, all the examples I saw set the type, so I did the same to not change the behavior. But it would be useful to check and clarify whether this field is required or not. I don't think it is used. It can lead to inconsistent models (where the actual attribute is of type int and the reference says float etc., which we don't check for, I think.) It is simpler to pick up the type from the attribute passed into the function.

Reply:
I don't remember what was the exact issue this was at least 2 years back and since then we have made quite a bit of improvements in this area.

I agree adding a type here can lead to inconsistent models. Can we move to picking the type from attr instead?

Reply:
Not sure I understood "pick the type from attr instead". I think we are doing that already. I think we just need to clarify in the spec that setting the type for this case in not needed (and that this type is ignored). And decide what to do in the checker. And update the current uses to drop this. But the issue is compatibility, if other users have custom functions that do this. In any case, I think this is a separate discussion, and we should consider it separately from the current PR.

Reply:
I tried removing the " : ints". However, the checker is currently checking this type, so this causes a failure. I am rolling back that change, and adding the ": ints" back.